pkg/controller/utils/utils.go

```diff
@@ -187,19 +200,59 @@ func IsPVCTerminating(pvc *corev1.PersistentVolumeClaim) bool {
 	return pvc.DeletionTimestamp != nil
 }
 
+// GetDesiredImage returns the desired image for the input containerName from the aeroCluster spec.
+func GetDesiredImage(aeroCluster *aerospikev1alpha1.AerospikeCluster, containerName string) (string, error) {
+	if containerName == "aerospike-server" {
+		return aeroCluster.Spec.Build, nil
+	}
+
+	for _, sidecar := range aeroCluster.Spec.PodSpec.Sidecars {
+		if sidecar.Name == containerName {
+			return sidecar.Image, nil
+		}
+	}
+
+	return "", fmt.Errorf("Container %s not found", containerName)
+}
+
 // IsPodUpgraded assume that all container have same image or take containerID
-func IsPodUpgraded(pod *corev1.Pod, image string) bool {
+func IsPodUpgraded(pod *corev1.Pod, aeroCluster *aerospikev1alpha1.AerospikeCluster) bool {
 	pkglog.Info("Checking pod image")
 
 	if !IsPodRunningAndReady(pod) {
 		return false
 	}
 
-	for _, ps := range pod.Status.ContainerStatuses {
-		if !ps.Ready || !IsImageEqual(ps.Image, image) {
+	return IsPodOnDesiredImage(pod, aeroCluster)
+}
+
+// getPodContainerStatus provides status for container in a pod if present else nil.
+// Abstracted into a function because in come cases pod.Status.ContainerStatuses was empty.
+func getPodContainerStatus(pod *corev1.Pod, containerName string) *corev1.ContainerStatus {
+	for _, status := range pod.Status.ContainerStatuses {
+		if status.Name == containerName {
+			return &status
+		}
+	}
+
+	return nil
+}
+
+// IsPodOnDesiredImage indicates of pod is ready and on desired images for all containers.
+func IsPodOnDesiredImage(pod *corev1.Pod, aeroCluster *aerospikev1alpha1.AerospikeCluster) bool {
+	for _, ps := range pod.Spec.Containers {
+		desiredImage, err := GetDesiredImage(aeroCluster, ps.Name)
+		if err != nil {
+			// Maybe a deleted sidecar. Ignore.
```

Comment:
If container is deleted then should we check for it's status also or just go to the next container?

Reply:
This is the case where the user has removed the sidecar from podSpec. Since he has removed the sidecar we can assume that we are on the desired Image, since this container will eventually be removed during a rolling upgrade.